src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,194 @@
+internal partial class RazorProjectInfoEndpointPublisher : IDisposable
+{
+    private readonly LSPRequestInvoker _requestInvoker;
+    private readonly IProjectSnapshotManager _projectManager;
+
+    private readonly AsyncBatchingWorkQueue<(IProjectSnapshot Project, bool Removal)> _workQueue;
+    private readonly CancellationTokenSource _disposeTokenSource;
+
+    // Delay between publishes to prevent bursts of changes yet still be responsive to changes.
+    private static readonly TimeSpan s_enqueueDelay = TimeSpan.FromMilliseconds(250);
+
+    [ImportingConstructor]
+    public RazorProjectInfoEndpointPublisher(
+        LSPRequestInvoker requestInvoker,
+        IProjectSnapshotManager projectManager)
+        : this(requestInvoker, projectManager, s_enqueueDelay)
+    {
+    }
+
+    // Provided for tests to specify enqueue delay
+    public RazorProjectInfoEndpointPublisher(
+        LSPRequestInvoker requestInvoker,
+        IProjectSnapshotManager projectManager,
+        TimeSpan enqueueDelay)
+    {
+        _requestInvoker = requestInvoker;
+        _projectManager = projectManager;
+
+        _disposeTokenSource = new();
+        _workQueue = new(
+            enqueueDelay,
+            ProcessBatchAsync,
+            _disposeTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    public void StartSending()
+    {
+        _projectManager.Changed += ProjectManager_Changed;
+
+        var projects = _projectManager.GetProjects();
+        foreach (var project in projects)
+        {
+            // Don't enqueue project addition as those are unlikely to come through in large batches.
+            // Also ensures that we won't get project removal go through the queue without project addition.
+            ImmediatePublish(project, _disposeTokenSource.Token);
+        }
+    }
+
+    private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
+    {
+        // Don't do any work if the solution is closing
+        if (args.SolutionIsClosing)
+        {
+            return;
+        }
+
+        switch (args.Kind)
+        {
+            case ProjectChangeKind.ProjectChanged:
+            case ProjectChangeKind.DocumentRemoved:
+            case ProjectChangeKind.DocumentAdded:
+                if (args.Newer is null)
+                {
+                    break;
+                }
+
+                EnqueuePublish(args.Newer.AssumeNotNull());
+
+                break;
+
+            case ProjectChangeKind.ProjectAdded:
+
+                if (args.Newer is not null)
+                {
+                    // Don't enqueue project addition as those are unlikely to come through in large batches.
+                    // Also ensures that we won't get project removal go through the queue without project addition.
+                    ImmediatePublish(args.Newer.AssumeNotNull(), _disposeTokenSource.Token);
```

Comment:
Nit: I think (hope) `.AssumeNotNull()` is not needed now that there is an actual null check

Reply:
Oops, had this set as automerge, sorry - will need to do a small clean-up PR anyway to address this, to proper serialization for the params and add unit tests.

Reply:
Yeah, I was very surprised this auto-merged after my approval 😛

Reply:
I'm surprised this was marked as auto-merge too, since I didn't get an opportunity to look after my feedback had been addressed.